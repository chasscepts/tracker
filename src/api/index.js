import axios from 'axios';

const normalizeError = (err) => {
  if (!err) {
    return { message: 'An unknown error encountered. Please try again.' };
  }
  if (err.response) {
    return { message: err.response.data.message || JSON.stringify(err.response.data) };
  }
  if (err.request) {
    return { message: 'Server is not responding.' };
  }
  if (err.message) {
    return { message: err.message };
  }
  if (typeof err === 'string') {
    return { message: err };
  }
  return { message: 'An unknown error encountered. Please try again.' };
};

const instantiate = (token) => {
  const instance = axios.create({
    baseURL: 'http://localhost:3000/',
    responseType: 'json',
  });
  instance.defaults.headers.common.Authorization = `Bearer ${token}`;

  return instance;
};

/**
 *
 * @param {axios} instance An instance of axios to use for this request
 * @param {string} path relative url
 * @returns Promise that resolves to fetched data when request is successful
 * and rejects with error when request fails
 */
const get = (instance, path) => new Promise((resolve, reject) => {
  instance.get(path)
    .then((res) => resolve(res.data))
    .catch((err) => reject(normalizeError(err)));
});

/**
 *
 * @param {axios} instance An instance of axios to use for this request
 * @param {string} path relative url
 * @param {{ string: any }} data sent in body of this post
 * @returns Promise that resolves to fetched data when request is successful
 * and rejects with error when request fails
 */
const post = (instance, path, data) => new Promise((resolve, reject) => {
  instance.post(path, data)
    .then((res) => resolve(res.data))
    .catch((err) => reject(normalizeError(err)));
});

const api = {
  getGroups: (token) => get(instantiate(token), '/groups'),
  getTasks: (token, start, end) => {
    let path = '/tasks';
    if (start) {
      path = `${path}?start=${start}`;
      if (end) {
        path = `${path}&end=${end}`;
      }
    }

    return get(instantiate(token), path);
  },
  updateTask: (token, title) => post(instantiate(token), '/tasks', { title }),
};

export default api;
