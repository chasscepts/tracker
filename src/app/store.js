import { configureStore } from '@reduxjs/toolkit';
import authReducer from '../reducers/authSlice';
import tasksReducer from '../reducers/tasksSlice';
import timerReducer from '../reducers/timerSlice';
import detailsReducer from '../reducers/detailsSlice';
import feedbackReducer from '../reducers/feedbackSlice';

export const createStore = () => configureStore({
  reducer: {
    auth: authReducer,
    tasks: tasksReducer,
    timer: timerReducer,
    details: detailsReducer,
    feedback: feedbackReducer,
  },
});

const store = createStore();

export default store;
