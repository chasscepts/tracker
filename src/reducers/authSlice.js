import { createSlice } from '@reduxjs/toolkit';
import axios from 'axios';

/* eslint-disable no-param-reassign */
const authSlice = createSlice({
  name: 'auth',
  initialState: {
    authenticated: false,
    loginError: null,
    registrationError: null,
    token: null,
    registrationSuccess: false,
  },
  reducers: {
    setToken: (state, { payload }) => {
      state.authenticated = true;
      state.token = payload;
      state.loginError = null;
      state.registrationError = null;
      state.registrationSuccess = false;
    },
    setLoginError: (state, { payload }) => {
      state.authenticated = false;
      state.token = null;
      state.loginError = payload;
      state.registrationError = null;
      state.registrationSuccess = false;
    },
    setRegistrationError: (state, { payload }) => {
      state.authenticated = false;
      state.token = null;
      state.loginError = null;
      state.registrationError = payload;
      state.registrationSuccess = false;
    },
    setRegistrationSuccess: (state) => {
      state.authenticated = false;
      state.token = null;
      state.loginError = null;
      state.registrationError = null;
      state.registrationSuccess = true;
    },
    logout: (state) => {
      state.authenticated = false;
      state.token = null;
      state.loginError = null;
      state.registrationError = null;
      state.registrationSuccess = false;
    },
  },
});
/* eslint-enable no-param-reassign */

export const {
  setToken,
  setLoginError,
  setRegistrationError,
  setRegistrationSuccess,
  logout,
} = authSlice.actions;

export const login = (email, password) => (dispatch) => {
  const url = 'http://localhost/auth';
  axios({
    url,
    method: 'post',
    responseType: 'json',
    data: { email, password },
  })
    .then((res) => {
      dispatch(setToken(res.data.token));
    })
    .catch((err) => {
      if (!err) {
        dispatch(setLoginError('Unknown error encountered. Please try again.'));
        return;
      }
      if (err.response) {
        dispatch(setLoginError(err.response.data));
        return;
      }
      if (err.request) {
        dispatch(setLoginError('Server is not responding.'));
        return;
      }
      if (err.message) {
        dispatch(setLoginError(err.message));
        return;
      }
      if (typeof err === 'string') {
        dispatch(setLoginError(err));
        return;
      }
      dispatch(setLoginError('Unknown error encountered. Please try again.'));
    })
    .catch((err) => {
      console.log(err);
    });
};

export const register = (email, password) => (dispatch) => {
  const url = 'http://localhost/auth/register';
  axios({
    url,
    method: 'post',
    responseType: 'json',
    data: { email, password },
  })
    .then(() => {
      dispatch(setRegistrationSuccess());
    })
    .catch((err) => {
      if (!err) {
        dispatch(setRegistrationError('An unknown error encountered. Please try again.'));
        return;
      }
      if (err.response) {
        dispatch(setRegistrationError(err.response.data));
        return;
      }
      if (err.request) {
        dispatch(setRegistrationError('Server is not responding.'));
        return;
      }
      if (err.message) {
        dispatch(setRegistrationError(err.message));
        return;
      }
      if (typeof err === 'string') {
        dispatch(setRegistrationError(err));
        return;
      }
      dispatch(setRegistrationError('An unknown error encountered. Please try again.'));
    })
    .catch((err) => {
      console.log(err);
    });
};

export const selectAuthenticated = (state) => state.auth.authenticated;

export const selectLoginError = (state) => state.auth.loginError;

export const selectRegistrationError = (state) => state.auth.registrationError;

export default authSlice.reducer;
