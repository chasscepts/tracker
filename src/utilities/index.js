const doubleDigits = (num) => {
  if (num < 10) return `0${num}`;
  return num;
};

/**
 * @param {Date} date  Date object
 * @returns string representation of date that conforms to API format
 */
const formatDate = (date) => `${date.getFullYear()}-${doubleDigits(date.getMonth() + 1)}-${doubleDigits(date.getDate())}`;

export const COLORS = {
  primary: '#62b5e5',
  gray: '#67899c',
  red: '#9e0606',
};

export const dates = {
  /**
   * @returns string resentation of today that conforms to API format
   */
  today: () => formatDate(new Date()),
  format: formatDate,
  timeStringClock: (time) => {
    let rem = time;
    const hrs = Math.floor(rem / 3600);
    rem -= 3600 * hrs;
    const mins = Math.floor(rem / 60);
    const secs = Math.floor(rem - (60 * mins));
    return `${doubleDigits(hrs)} : ${doubleDigits(mins)} : ${doubleDigits(secs)}`;
  },
  timeString: (time) => {
    let rem = time;
    const hrs = Math.floor(rem / 3600);
    rem -= 3600 * hrs;
    const mins = Math.floor(rem / 60);
    const secs = Math.floor(rem - (60 * mins));
    const ch = hrs === 1 ? 'hr' : 'hrs';
    const cm = mins === 1 ? 'min' : 'mins';
    const cs = secs === 1 ? 'sec' : 'secs';
    return `${hrs}${ch} ${mins}${cm} ${secs}${cs}`;
  },
  toDate: (str) => {
    const parts = str.split('-');
    return new Date(parts[0], parts[1] - 1, parts[2]);
  },
};

export const taskDuration = (task, date) => {
  const entry = task.entries.find((entry) => entry.entry_date === date);
  return (entry && entry.duration) || 0;
};
