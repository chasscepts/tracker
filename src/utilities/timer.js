let elapsed = 0;
let seconds = 0;
let stopped = true;
let prevStamp;
let client;

let id = 0;

let listeners = [];

const subscribe = (callback) => {
  const temp = id;
  id += 1;
  listeners.push({ id, callback });
  return temp;
};

const unsubscribe = (id) => {
  listeners = listeners.filter((l) => l.id !== id);
};

const raiseSecondElapsed = (elapsed) => {
  listeners.forEach((l) => l.callback(elapsed));
};

const loop = (timestamp) => {
  if (stopped) return;
  if (prevStamp === undefined) {
    prevStamp = timestamp;
  } else {
    elapsed += timestamp - prevStamp;
    prevStamp = timestamp;
    const temp = Math.floor(elapsed / 1000);
    if (temp !== seconds) {
      seconds = temp;
      raiseSecondElapsed(temp);
    }
  }
  requestAnimationFrame(loop);
};

const stop = () => {
  stopped = true;
  prevStamp = undefined;
  client = null;
};

const start = (params) => {
  client = params;
  elapsed = 0;
  seconds = 0;
  stopped = false;
  requestAnimationFrame(loop);
};

const resume = () => {
  stopped = false;
  requestAnimationFrame(loop);
};

export default {
  start,
  stop,
  resume,
  subscribe,
  unsubscribe,
  running: () => !stopped,
  elapsedTime: () => seconds,
  getClient: () => client,
};
