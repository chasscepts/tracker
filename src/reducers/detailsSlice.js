import { createSlice } from '@reduxjs/toolkit';
import api from '../api';

/* eslint-disable no-param-reassign */
const detailsSlice = createSlice({
  name: 'details',
  initialState: {
    requestCount: 0,
    hasPendingRequest: false,
    store: [],
    group: null,
    loadError: null,
  },
  reducers: {
    addRequest: (state) => {
      state.requestCount += 1;
      state.hasPendingRequest = true;
    },
    addGroup: (state, { payload }) => {
      state.group = payload;
      state.store.push(payload);
      let temp = state.requestCount - 1;
      if (temp < 0) {
        temp = 0;
        state.hasPendingRequest = false;
      }
      state.requestCount = temp;
    },
    setGroup: (state, { payload }) => {
      state.group = payload;
      let temp = state.requestCount - 1;
      if (temp < 0) {
        temp = 0;
        state.hasPendingRequest = false;
      }
      state.requestCount = temp;
    },
    setGroupLoadError: (state, { payload }) => {
      state.loadError = payload;
    },
  },
});
/* eslint-enable no-param-reassign */

export const {
  addRequest,
  addGroup,
  setGroup,
  setGroupLoadError,
} = detailsSlice.actions;

export const selectDetailsHasPendingError = (state) => state.details.hasPendingRequest;
export const selectDetailsGroup = (state) => state.details.group;
export const selectDetailsLoadError = (state) => state.details.loadError;

export const loadGroupTasks = (groupId, title) => (dispatch, getState) => {
  const { auth: { user }, details: { store } } = getState();
  if (!user) {
    dispatch(setGroupLoadError('You must be logged in to fetch tasks'));
    return;
  }

  const group = store.find((g) => g.id === groupId);
  if (group) {
    dispatch(setGroup(group));
    return;
  }

  dispatch(addRequest());

  api.getGroupTasks(user.token, groupId)
    .then((group) => {
      dispatch(addGroup({ title, id: groupId, tasks: group }));
    })
    .catch((err) => setGroupLoadError(err.message));
};

export default detailsSlice.reducer;
