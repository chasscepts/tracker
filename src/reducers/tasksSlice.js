import { createSlice } from '@reduxjs/toolkit';
import {
  fetchGroups, fetchTasks, createTask, updateTask, deleteTask,
} from '../api';
import { today, daysAgo } from '../utilities/dates';
import { addFeedback } from './feedbackSlice';

/* eslint-disable no-param-reassign */
const tasksSlice = createSlice({
  name: 'tasks',
  initialState: {
    date: today(),
    groups: null,
    groupsLoading: false,
    tasksCache: {},
    tasks: [],
    tasksLoading: false,
    createRequestCount: 0,
    updateRequestCount: 0,
  },
  reducers: {
    setDate: (state, { payload }) => {
      state.date = payload;
    },
    setGroups: (state, { payload }) => {
      state.groups = payload;
      state.groupsLoading = false;
    },
    setGroupsLoading: (state, { payload }) => {
      state.groupsLoading = payload;
    },
    setTasks: (state, { payload }) => {
      state.tasks = payload;
    },
    setTasksLoading: (state, { payload }) => {
      state.tasksLoading = payload;
    },
    updateTasksCache: (state, { payload: { date, tasks } }) => {
      state.tasksCache[date] = tasks;
    },
    addCreateTaskRequest: (state) => {
      state.createRequestCount += 1;
    },
    addTask: (state, { payload }) => {
      state.tasks.push(payload);
      if (state.createRequestCount > 0) {
        state.createRequestCount -= 1;
      }
    },
    removeCreateRequest: (state) => {
      if (state.createRequestCount > 0) {
        state.createRequestCount -= 1;
      }
    },
    addUpdateTaskRequest: (state) => {
      state.updateRequestCount += 1;
    },
    removeUpdateTaskRequest: (state) => {
      if (state.updateRequestCount > 0) {
        state.updateRequestCount -= 1;
      }
    },
    updateLocalTask: (state, { payload: { id, title } }) => {
      const task = state.tasks.find((t) => t.id === id);
      if (task) {
        task.title = title;
      }
      Object.keys(state.tasksCache).forEach((key) => {
        const task = state.tasksCache[key].find((t) => t.id === id);
        if (task) {
          task.title = title;
        }
      });
      if (state.updateRequestCount > 0) {
        state.updateRequestCount -= 1;
      }
    },
    deleteLocalTask: (state, { payload }) => {
      state.tasks = state.tasks.filter((t) => t.id !== payload);
      if (state.updateRequestCount > 0) {
        state.updateRequestCount -= 1;
      }
    },
    updateTaskEntry: (state, { payload: { entry, duration } }) => {
      const task = state.tasks.find((t) => t.id === entry.task_id);
      if (task) {
        task.entries[0].duration += duration;
      }
    },
  },
});
/* eslint-enable no-param-reassign */

export const {
  setDate,
  setGroups,
  setGroupsLoading,
  setTasks,
  setTasksLoading,
  updateTasksCache,
  addTask,
  addCreateTaskRequest,
  removeCreateRequest,
  updateLocalTask,
  addUpdateTaskRequest,
  removeUpdateTaskRequest,
  deleteLocalTask,
  updateTaskEntry,
} = tasksSlice.actions;

export const selectDate = (state) => state.tasks.date;

export const selectGroups = (state) => state.tasks.groups;

export const selectGroupsLoading = (state) => state.tasks.groupsLoading;

export const selectTasks = (state) => state.tasks.tasks;

export const selectCreateTaskRequestCount = (state) => state.tasks.createRequestCount;

export const selectUpdateTaskRequestCount = (state) => state.tasks.updateRequestCount;

export const loadGroups = () => (dispatch, getState) => {
  const state = getState();
  if (state.tasks.groups || state.tasks.groupsLoading) return;

  const { auth: { user } } = state;

  if (!user) {
    dispatch(addFeedback({ message: 'You must be logged in to fetch groups', type: 'error' }));
    return;
  }

  dispatch(setGroupsLoading(true));

  fetchGroups(user.token)
    .then((groups) => {
      dispatch(setGroups(groups));
    })
    .catch(({ message }) => {
      dispatch(setGroupsLoading(false));
      dispatch(addFeedback({ message, type: 'error' }));
    });
};

export const loadTasks = () => (dispatch, getState) => {
  const state = getState();

  const { auth: { user } } = state;

  if (!user) {
    dispatch(addFeedback({ message: 'You must be logged in to fetch tasks!', type: 'error' }));
    return;
  }

  dispatch(setTasksLoading());

  fetchTasks(user.token, { start: daysAgo(14) })
    .then((tasks) => {
      dispatch(setTasks(tasks));
    })
    .catch(({ message }) => dispatch(addFeedback({ message, type: 'error' })));
};

export const createTaskAsync = (groupId, title) => (dispatch, getState) => {
  const state = getState();
  const { auth: { user } } = state;

  if (!user) {
    dispatch(addFeedback({ message: 'You must be logged in to create tasks', type: 'error' }));
    return;
  }

  dispatch(addCreateTaskRequest());

  createTask(user.token, groupId, title)
    .then((task) => {
      dispatch(addTask(task));
      dispatch(addFeedback({ message: `The task ${task.title} was successfully created.`, type: 'success' }));
    })
    .catch(({ message }) => {
      dispatch(removeCreateRequest());
      dispatch(addFeedback({ message, type: 'error' }));
    });
};

export const updateTaskAsync = (id, title) => (dispatch, getState) => {
  const state = getState();
  const { auth: { user } } = state;

  if (!user) {
    dispatch(addFeedback({ message: 'You must be logged in to update tasks', type: 'error' }));
    return;
  }

  dispatch(addUpdateTaskRequest());

  updateTask(user.token, id, title)
    .then(() => {
      dispatch(updateLocalTask({ id, title }));
      dispatch(addFeedback({ message: `Task ${title} successfully updated.`, type: 'success' }));
    })
    .catch(({ message }) => {
      dispatch(removeUpdateTaskRequest());
      dispatch(addFeedback({ message, type: 'error' }));
    });
};

export const deleteTaskAsync = (id) => (dispatch, getState) => {
  const state = getState();
  const { auth: { user } } = state;

  if (!user) {
    dispatch(addFeedback({ message: 'You must be logged in to update tasks', type: 'error' }));
    return;
  }

  dispatch(addUpdateTaskRequest());

  deleteTask(user.token, id)
    .then(() => {
      dispatch(deleteLocalTask(id));
      dispatch(addFeedback({ message: 'Task successfully deleted.', type: 'success' }));
    })
    .catch(({ message }) => {
      dispatch(removeUpdateTaskRequest());
      dispatch(addFeedback({ message, type: 'error' }));
    });
};

export default tasksSlice.reducer;
