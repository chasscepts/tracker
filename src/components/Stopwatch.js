import { useState, useEffect } from 'react';
import { Redirect } from 'react-router';
import { useDispatch, useSelector } from 'react-redux';
import PropTypes from 'prop-types';
import css from '../assets/css/Stopwatch.module.css';
import btn from '../assets/css/button.module.css';
import sphere from '../assets/images/sphere.png';
import timer from '../utilities/timer';
import {
  commitNext, selectEntry, selectEntryPendingRequestCount, selectEntryUpdateError,
  selectEntryUpdateSuccess, selectNextEntry, updateEntry,
} from '../reducers/timerSlice';
import LdsRing from './LdsRing';
import { dates } from '../utilities';

const styles = {
  ml20: {
    marginLeft: '20px',
  },
};

function Display({ time }) {
  return <div className={css.time}>{dates.timeStringClock(time)}</div>;
}

Display.propTypes = {
  time: PropTypes.number,
};

Display.defaultProps = {
  time: 0,
};

const cocurrentMsg = `Timer is already started for another task.
Cocurrent running of tasks is not yet supported.
Stop timer to load a new session.`;

function MessageBoard({ msg, type }) {
  if (!msg) return <></>;
  return (
    <div className={css[type]}>{msg}</div>
  );
}

MessageBoard.propTypes = {
  msg: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.bool,
  ]),
  type: PropTypes.string,
};

MessageBoard.defaultProps = {
  msg: false,
  type: 'info',
};

function UpdatingIndicator({ count }) {
  if (count === 0) return <></>;
  const entry = count === 1 ? 'Entry' : 'Entries';

  return (
    <div className={css.loader}>
      <LdsRing width={30} color="#62b5e5" />
      <div className={css.loaderText}>
        {`Updating ${count} ${entry}`}
      </div>
    </div>
  );
}

UpdatingIndicator.propTypes = {
  count: PropTypes.number.isRequired,
};

export default function Stopwatch() {
  const [elapsed, setElapsed] = useState(0);
  const [running, setRunning] = useState(timer.running());
  const { entry, title } = useSelector(selectEntry);
  const { nextEntry } = useSelector(selectNextEntry);
  const successMsg = useSelector(selectEntryUpdateSuccess);
  const errorMsg = useSelector(selectEntryUpdateError);
  const requestCount = useSelector(selectEntryPendingRequestCount);
  const dispatch = useDispatch();

  if (!entry) {
    return <Redirect to="/" />;
  }

  if (!running && nextEntry && nextEntry.id !== entry.id) {
    dispatch(commitNext());
    return <></>;
  }

  const hasNext = (running && nextEntry && entry.id !== nextEntry.id) || false;

  const btnClass = (color) => `${btn.btn} ${btn[color]} ${btn.round}`;

  const handClass = running ? `${css.hand} ${css.active}` : css.hand;

  const timerId = timer.subscribe((seconds) => {
    setElapsed(seconds);
  });

  const startTimer = () => {
    timer.start();
    setRunning(true);
  };

  const commit = () => {
    timer.stop();
    setRunning(false);
    dispatch(updateEntry(entry, timer.elapsedTime(), title));
    if (hasNext) {
      dispatch(commitNext());
    }
  };

  const discard = () => {
    timer.stop();
    setRunning(false);
    setElapsed(0);
  };

  useEffect(() => () => timer.unsubscribe(timerId));

  return (
    <div className={css.container}>
      <div className={css.stopwatch}>
        <div className={css.handWrap}>
          <div className={handClass}>
            <img className={css.sphere} src={sphere} alt="sphere" />
          </div>
        </div>
        <Display time={elapsed} />
      </div>
      <UpdatingIndicator count={requestCount} />
      <div className={css.entryWrap}>
        {`${title}: ${dates.timeString(entry.duration)} recorded`}
      </div>
      <MessageBoard msg={successMsg} type="success" />
      <MessageBoard msg={errorMsg} type="error" />
      <MessageBoard msg={cocurrentMsg && hasNext} type="error" />
      <div className={css.controls}>
        {running && (
        <>
          <button className={btnClass('blue')} type="button" onClick={commit}>Commit Record</button>
          <button className={btnClass('red')} type="button" onClick={discard} style={styles.ml20}>Discard</button>
        </>
        )}
        {!running && <button className={btnClass('blue')} type="button" onClick={startTimer}>Start Task</button>}
      </div>
    </div>
  );
}
