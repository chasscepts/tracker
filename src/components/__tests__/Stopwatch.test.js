import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen, waitFor } from '@testing-library/react';
import Wrapper from '../../test_helpers/wrapper';
import Stopwatch from '../Stopwatch';
import { updateEntry } from '../../api';
import { createStore } from '../../app/store';
import login from '../../test_helpers/login';
import { setTasks } from '../../reducers/tasksSlice';
import { subscribe, elapsedTime } from '../../utilities/timer';
import Router from 'react-router';
import userEvent from '@testing-library/user-event';

jest.mock('../../utilities/timer', () => ({
  ...jest.requireActual('../../utilities/timer'),
  subscribe: jest.fn(),
  elapsedTime: jest.fn(),
}));

jest.mock('../../api');

const tasks = [
  {
    id: 1,
    title: 'Task 1',
    group_id: 1,
    user_id: 1,
    entries: [
      {
        id: 1, task_id: 1, entry_date: '2021-09-21', duration: 0,
      },
    ],
  },
];

describe('Stopwatch', () => {
  it('matches snapshot', () => {
    const store = createStore();
    login(store);
    const tree = renderer.create(<Wrapper Component={Stopwatch} store={store} />);
    expect(tree.toJSON()).toMatchSnapshot();
    tree.unmount();
  });

  it('displays a "select task" prompt', () => {
    const store = createStore();
    login(store); 
    render(<Wrapper Component={Stopwatch} store={store} />);
    expect(screen.getByText('Please Select a Task')).toBeInTheDocument();
  });

  it('does not display start button when entry is not selected', () => {
    const store = createStore();
    login(store); 
    render(<Wrapper Component={Stopwatch} store={store} />);
    expect(screen.queryByRole('button', { name: 'Start Task' })).not.toBeInTheDocument();
  });

  it('display start button when entry is selected', () => {
    const store = createStore();
    login(store);
    store.dispatch(setTasks(tasks));
    const spy = jest.spyOn(Router, 'useParams').mockReturnValue({ id: 1, taskId: 1 });
    render(<Wrapper Component={Stopwatch} store={store} />);
    expect(screen.getByRole('button', { name: 'Start Task' })).toBeInTheDocument();
    spy.mockClear();
  });

  it('subscribes to timer', async () => {
    const store = createStore();
    await login(store);
    subscribe.mockReturnValue(1);
    const spy = jest.spyOn(Router, 'useParams').mockReturnValue({ });
    render(<Wrapper Component={Stopwatch} store={store} />);
    expect(subscribe).toHaveBeenCalled();
    spy.mockClear();
  });

  it('calls to api to record entry', async () => {
    const store = createStore();
    login(store);
    store.dispatch(setTasks(tasks));
    const spy = jest.spyOn(Router, 'useParams').mockReturnValue({ id: 1, taskId: 1 });
    updateEntry.mockResolvedValue('');
    subscribe.mockReturnValue(1);
    elapsedTime.mockReturnValue(100);
    render(<Wrapper Component={Stopwatch} store={store} />);
    userEvent.click(screen.getByRole('button', { name: 'Start Task' }));
    await waitFor(() => screen.getByRole('button', { name: 'Commit Record' }));
    userEvent.click(screen.getByRole('button', { name: 'Commit Record' }));
    expect(updateEntry).toHaveBeenCalled();
    spy.mockClear();
  });
});
