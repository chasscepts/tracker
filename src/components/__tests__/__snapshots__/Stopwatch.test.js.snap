// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stopwatch matches snapshot 1`] = `
<div
  className="container"
>
  <div
    className="stopwatch"
  >
    <div
      className="handWrap"
    >
      <div
        className="hand"
      >
        <img
          alt="sphere"
          className="sphere"
          src="sphere.png"
        />
      </div>
    </div>
    <div
      className="time"
    >
      00 : 00 : 00
    </div>
  </div>
  <div
    className="entryWrap"
  >
    No Task Selected
  </div>
  <div
    className="controls"
  >
    <div
      className="notSelected"
    >
      Please Select a Task
    </div>
  </div>
</div>
`;
